content/en/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,286 @@
+
+**Note**: If a container has no memory limits specified, we substitute `limits.memory` for node allocatable memory.
+
+**Summary:**
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.max</td>
+        <td><code>memory.max</code> specifies the maximum memory limit,
+        a container is allowed to use. If a container tries to consume
+        more memory than the configured limit, kernel terminates the 
+        process with an Out of Memory (OOM) error. 
+        <br>
+        <br>
+        <i>It is mapped to the container's memory limit specified in Pod manifest.</i>
+        </td>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory 
+        the cgroups must always retain, i.e., memory that can never be
+        reclaimed by the system.
+        If there's no unprotected reclaimable memory available, OOM kill is invoked.
+        <br>
+        <br>
+        <i>It is mapped to the container's memory request specified in Pod manifest..</i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> specifies the memory usage throttle limit. 
+       This is the main mechanism to control a cgroup's memory use. If 
+       cgroups memory use goes over the high boundary specified here,
+       the cgroups processes are throttled and put under heavy reclaim pressure.
+       <br>
+       <br>
+       <i>Kubernetes uses a formula to calculate <code>memory.high</code>,
+       depending on container's memory request, memory limit or node allocatable memory
+      (if container's memory limit is empty) and a throttling factor.
+      Please refer to the <a href="https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos">KEP</a>
+      for more details on the formula.</i>
+       </td>
+   </tr>
+</table>
+
+**Note** `memory.high` is set only on container level cgroups while `memory.min` is set on
+container, pod and node level cgroups.
+
+### `memory.min` calculations for cgroups heirarchy
+
+When container memory requests are made, kubelet passes `memory.min` to the back-end 
+CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 
+container creation. The `memory.min` in container level cgroups will be set to:  
+
+$memory.min =  pod.spec.containers[i].resources.requests[memory]$  
+<sub>for every i<sup>th</sup> container in a pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroups directories are all 
+set, the pod and node cgroups directories need to be set correctly. 
+
+`memory.min` in pod level cgroup:  
+$memory.min = \sum_{i=0}^{no. of pods}pod.spec.containers[i].resources.requests[memory]$  
+
+`memory.min` in node level cgroup:  
+$memory.min = \sum_{i}^{no. of nodes}\sum_{j}^{no. of pods}pod[i].spec.containers[j].resources.requests[memory]$  
```

Comment:
This might be easier to read as an SVG formula with alt text.

Reply:
looks fine to me in the preview. Also, it will make it look more consistent to either use SVG for all, or use markdown foll all. Attaching ss. Thoughts? 
![Screenshot 2023-04-17 at 11 49 29 AM](https://user-images.githubusercontent.com/5120675/232582889-b1898271-d0d8-49e7-afe9-144bd2945eb3.png)